File: core/game_record.cpp
#include "game_record.h"
#include "game.h"

TablebaseFirstHitStats tablebase_first_hit_stats(const GameRecord &record) {
  TablebaseFirstHitStats s{};
  int last_tb[2] = {-1, -1};
  for (const auto &turn : record.turns()) {
    int p = turn.current_player;
    if (turn.tb_case == 1) {
      if (last_tb[p] != 1) {
        ++s.case1;
        s.case1_seq_sum += turn.tb_forced_seq_len;
        if (turn.tb_forced_seq_len > s.case1_seq_max)
          s.case1_seq_max = turn.tb_forced_seq_len;
      }
      last_tb[p] = 1;
    } else if (turn.tb_case == 2) {
      if (last_tb[p] != 2) {
        ++s.case2;
        if (turn.tb_opp1_table_straight)
          ++s.case2_table_straight;
      }
      last_tb[p] = 2;
    } else {
      last_tb[p] = -1;
    }
  }
  return s;
}

GameRecord::GameRecord() {}

void GameRecord::set_initial_state(const Game &game) {
  _game = game;
  _views[0] = PartialGame(game, 0);
  _views[1] = PartialGame(game, 1);
}

void GameRecord::add_move(const Move &move, int tb_case, int tb_forced_seq_len,
                          bool tb_opp1_table_straight) {
  TurnRecord new_record{
      _game.current_player(),
      _game,
      _views,
      _views[_game.current_player()].get_legal_moves(),
      _views[1 - _game.current_player()].get_possible_moves(),
      move,
      tb_case,
      tb_forced_seq_len,
      tb_opp1_table_straight};
  _turns.push_back(new_record);
  _game.apply_move(move);
  _views[0].apply_move(move);
  _views[1].apply_move(move);
}

File: crates/core/src/game_record.rs
use crate::game::Game;
use crate::moves::Move;
use crate::partial_game::PartialGame;

/// Everything recorded about a single turn for training / analysis.
pub struct TurnRecord {
    pub current_player: usize,
    pub game: Game,
    pub views: [PartialGame; 2],
    pub legal_moves: Vec<usize>,
    pub possible_moves: Vec<usize>,
    pub mv: Move,
    /// -1 = no tablebase, 1 = forced-win sequence, 2 = opponent has 1 card.
    pub tb_case: i8,
    pub tb_forced_seq_len: usize,
    pub tb_opp1_table_straight: bool,
}

#[derive(Default)]
pub struct TablebaseFirstHitStats {
    pub case1: u32,
    pub case2: u32,
    pub case1_seq_sum: u64,
    pub case1_seq_max: u32,
    pub case2_table_straight: u32,
}

pub struct GameRecord {
    initial_game: Game,
    views: [PartialGame; 2],
    pub turns: Vec<TurnRecord>,
}

impl GameRecord {
    pub fn new(game: &Game) -> Self {
        GameRecord {
            initial_game: game.clone(),
            views: [
                PartialGame::from_game(game, 0),
                PartialGame::from_game(game, 1),
            ],
            turns: Vec::new(),
        }
    }

    pub fn add_move(
        &mut self,
        mv: Move,
        game: &Game,
        tb_case: i8,
        tb_forced_seq_len: usize,
        tb_opp1_table_straight: bool,
    ) {
        let cp = game.current_player();
        let legal = self.views[cp].legal_moves();
        let possible = self.views[1 - cp].possible_moves();

        self.turns.push(TurnRecord {
            current_player: cp,
            game: game.clone(),
            views: self.views.clone(),
            legal_moves: legal,
            possible_moves: possible,
            mv,
            tb_case,
            tb_forced_seq_len,
            tb_opp1_table_straight,
        });

        self.views[0].apply_move(mv);
        self.views[1].apply_move(mv);
    }

    pub fn initial_game(&self) -> &Game {
        &self.initial_game
    }
}

pub fn tablebase_first_hit_stats(record: &GameRecord) -> TablebaseFirstHitStats {
    let mut stats = TablebaseFirstHitStats::default();
    let mut last_tb = [-1i8; 2];

    for turn in &record.turns {
        let p = turn.current_player;
        if turn.tb_case == 1 {
            if last_tb[p] != 1 {
                stats.case1 += 1;
                stats.case1_seq_sum += turn.tb_forced_seq_len as u64;
                if turn.tb_forced_seq_len as u32 > stats.case1_seq_max {
                    stats.case1_seq_max = turn.tb_forced_seq_len as u32;
                }
            }
            last_tb[p] = 1;
        } else if turn.tb_case == 2 {
            if last_tb[p] != 2 {
                stats.case2 += 1;
                if turn.tb_opp1_table_straight {
                    stats.case2_table_straight += 1;
                }
            }
            last_tb[p] = 2;
        } else {
            last_tb[p] = -1;
        }
    }
    stats
}
